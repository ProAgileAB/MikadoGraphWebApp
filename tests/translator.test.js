/* The responsiblity of the translator is to
   translate from the parse result data structure
   to graphviz text format
 */

const { parse } = require("../parser.js")
const { translate, translateElement } = require("../translator.js")

describe("mikado translator", () => {
    it("can translate a single goal", () => {
        expect(translate(parse("Goal\n  The goal\n"))).toMatchSnapshot()
    })
    it("gives a readable error if no goal is defined", () => {
        expect(translate(parse("Done: Goal\n"))).toMatchSnapshot()
    })
    it("handles single prerequisite", () => {
        expect(translate(parse("Goal\n" +
            "   A goal\n" +
            "Goal needs Format\n" +
            "   Format the file for readability\n"))).toMatchSnapshot()
    })
    it("handles single done prerequisite", () => {
        expect(translate(parse(`Goal
   A goal
Goal needs Format
   Format the file for readability
Done: Format
`))).toMatchSnapshot()
    })
})


describe("element translator", () => {
    it("points towards end node when Goal needs something", () => {
        const needElement = {
            text: "the text",
            needs: "Prereq",
            steps: ["Goal"]
        }
        expect(translateElement(needElement)).toEqual([
            `Prereq [label="the text"]`,
            `Prereq -> Goal [dir=back]`
        ])
    })
    it("points towards prerequisite node when general node needs something", () => {
        const needElement = {
            text: "This step needs to be done first",
            needs: "NeededPrereq",
            steps: ["PrereqInNeed"]
        }
        expect(translateElement(needElement)).toEqual([
            `NeededPrereq [label="This step needs to be done first"]`,
            `NeededPrereq -> PrereqInNeed [dir=back]`
        ])
    })
    describe("handles done by setting green background", () => {
        it("for goal", () => {
            const doneElement = { "done": "Goal" }
            expect(translateElement(doneElement)).toEqual([
                `Goal [style=filled, fillcolor=green]`
            ])
        });
        it("for non goal step", () => {
            const doneElement = { "done": "Step" }
            expect(translateElement(doneElement)).toEqual([
                `Step [style=filled, fillcolor=green]`
            ])
        });
    });
})
