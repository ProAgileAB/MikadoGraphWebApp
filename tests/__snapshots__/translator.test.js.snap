// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mikado translator can translate a single goal 1`] = `
"digraph G {
   Goal [peripheries=2, shape=oval, label=\\"The goal\\", style=filled, fillcolor=lightblue];

}"
`;

exports[`mikado translator gives a readable error if no goal is defined 1`] = `"Mikado graph needs exactly one Goal"`;

exports[`mikado translator handles single done prerequisite 1`] = `
"digraph G {
   Goal [peripheries=2, shape=oval, label=\\"A goal\\", style=filled, fillcolor=lightblue];
Format [label=\\"Format the file for readability\\"]
Format -> Goal [dir=back]
Format [style=filled, fillcolor=green]

}"
`;

exports[`mikado translator handles single prerequisite 1`] = `
"digraph G {
   Goal [peripheries=2, shape=oval, label=\\"A goal\\", style=filled, fillcolor=lightblue];
Format [label=\\"Format the file for readability\\"]
Format -> Goal [dir=back]

}"
`;
